File: src/org/lineageos/xiaomi_bluetooth/settings/configs/FindEarbudsController.java
package org.lineageos.xiaomi_bluetooth.settings.configs;

import android.content.Context;
import android.util.Log;

import androidx.annotation.NonNull;

import org.lineageos.xiaomi_bluetooth.R;
import org.lineageos.xiaomi_bluetooth.EarbudsConstants;
import org.lineageos.xiaomi_bluetooth.earbuds.Earbuds;
import org.lineageos.xiaomi_bluetooth.mma.MMADevice;

import java.io.IOException;


public class FindEarbudsController extends SwitchController {

    private static final String TAG = FindEarbudsController.class.getSimpleName();
    private static final boolean DEBUG = true;

    private static final byte LEFT_EARBUD_FLAG = 0x01;
    private static final byte RIGHT_EARBUD_FLAG = 0x02;
    private static final byte ENABLE_FLAG = 0x01;
    private static final byte DISABLE_FLAG = 0x00;

    private static final byte[] ENABLED_CONFIG = {ENABLE_FLAG, LEFT_EARBUD_FLAG | RIGHT_EARBUD_FLAG};
    private static final byte[] DISABLED_CONFIG = {DISABLE_FLAG, LEFT_EARBUD_FLAG | RIGHT_EARBUD_FLAG};

    public FindEarbudsController(@NonNull Context context, @NonNull String preferenceKey) {
        super(context, preferenceKey);
    }

    @NonNull
    @Override
    protected ConfigState getEnabledState() {
        return new ConfigState(ENABLED_CONFIG, R.string.find_earbuds_on);
    }

    @NonNull
    @Override
    protected ConfigState getDisabledState() {
        return new ConfigState(DISABLED_CONFIG, R.string.find_earbuds_off);
    }

    @Override
    public int getConfigId() {
        return EarbudsConstants.XIAOMI_MMA_CONFIG_FIND_EARBUDS;
    }

    @Override
    public int getExpectedConfigLength() {
        return 2;
    }

    @Override
    public boolean saveConfig(@NonNull MMADevice device, @NonNull Object value) throws IOException {
        if (!(value instanceof Boolean enable)) {
            Log.w(TAG, "Invalid value type: " + value.getClass().getSimpleName());
            return false;
        }
        Earbuds earbuds = validateEarbudsPresence(device);

        byte firstByte = enable ? ENABLE_FLAG : DISABLE_FLAG;
        byte secondByte = 0;
        if (earbuds.left != null) secondByte |= LEFT_EARBUD_FLAG;
        if (earbuds.right != null) secondByte |= RIGHT_EARBUD_FLAG;

        byte[] config = {firstByte, secondByte};
        return super.saveConfig(device, config);
    }

    @Override
    protected boolean isEnabled() {
        return isAvailable() && getConfigValue().length > 0 && getConfigValue()[0] == ENABLE_FLAG;
    }

    @NonNull
    private Earbuds validateEarbudsPresence(@NonNull MMADevice device) throws IOException {
        Earbuds earbuds = device.getBatteryInfo();

        if (earbuds == null || (earbuds.left == null && earbuds.right == null)) {
            throw new IllegalStateException("No earbuds information available");
        }

        return earbuds;
    }

}

File: src/org/lineageos/xiaomi_bluetooth/settings/configs/FindEarbudsController.kt
package org.lineageos.xiaomi_bluetooth.settings.configs

import android.content.Context
import org.lineageos.xiaomi_bluetooth.EarbudsConstants.XIAOMI_MMA_CONFIG_FIND_EARBUDS
import org.lineageos.xiaomi_bluetooth.R
import org.lineageos.xiaomi_bluetooth.mma.MMADevice

class FindEarbudsController(context: Context, preferenceKey: String) :
    SwitchController(context, preferenceKey) {

    override val enabledState = ConfigState(ENABLED_CONFIG, R.string.find_earbuds_on)
    override val disabledState = ConfigState(DISABLED_CONFIG, R.string.find_earbuds_off)

    override val configId = XIAOMI_MMA_CONFIG_FIND_EARBUDS
    override val expectedConfigLength = 2

    override val isEnabled
        get() = configValue?.get(0) == ENABLE_FLAG

    override fun saveConfig(device: MMADevice, value: Any): Boolean {
        require(value is Boolean) {
            "Invalid value type: ${value.javaClass.simpleName}"
        }

        val earbuds = device.batteryInfo
        check(earbuds.leftOrRightValid) {
            "Both earbuds disconnected"
        }

        val configValue = byteArrayOf(
            if (value) ENABLE_FLAG else DISABLE_FLAG,
            earbuds.run {
                val left = if (left.valid) LEFT_EARBUD_FLAG else 0
                val right = if (right.valid) RIGHT_EARBUD_FLAG else 0

                (left or right).toByte()
            }
        )

        return super.saveConfig(device, configValue)
    }

    companion object {
//        private val TAG = FindEarbudsController::class.java.simpleName
//        private const val DEBUG: Boolean = true

        private const val LEFT_EARBUD_FLAG = 0x01
        private const val RIGHT_EARBUD_FLAG = 0x02
        private const val ENABLE_FLAG: Byte = 0x01
        private const val DISABLE_FLAG: Byte = 0x00

        private val ENABLED_CONFIG = byteArrayOf(
            ENABLE_FLAG, (LEFT_EARBUD_FLAG or RIGHT_EARBUD_FLAG).toByte()
        )
        private val DISABLED_CONFIG = byteArrayOf(
            DISABLE_FLAG, (LEFT_EARBUD_FLAG or RIGHT_EARBUD_FLAG).toByte()
        )
    }
}
